Initial commit

autodidaddict opened the pull request:
This PR submits a proposal for an interface for delivering measurements from sensors.

autodidaddict commented:
Note that there are no shared libraries contained in this PR. That's deliberate - the responsibility of code gen should be in the consumers and not as part of a shared library that could create too much tight coupling and enforced lock-step release cadences.

thomastaylor312 commented:
This looks pretty nice. I'll defer to @ricochet as I don't do as much with sensors and IoT stuff

ricochet commented:
After doing some research, I found https://w3c.github.io/sensors/

Is this something you were aware of? It's new to me.

autodidaddict commented:
I've looked at it before. That interface is, if I recall, based on exposing sensors to JavaScript in a browser. The API interaction is less generalized than what I'm trying to accomplish here. Most of the difference is that there are string names for sensor types, sensor type events, and the JavaScript/browser has the ability to control when the sensor starts receiving, which this proposal doesn't do.

more than happy to amend this PR if there's something we think is worth copying/being inspired by in that W3C doc, but it feels like it's solving a slightly different problem.

autodidaddict commented:
It might also make sense to add a function that lets the component query the latest/current sensor value. This would help if the wasm component was producing aggregate calculations based on multiple sensors

autodidaddict commented:
Changed the sensor, network, and type IDs to strings.
Added an error handling function allowing a provider to deliver an error that may or may not correlate to a specific network or sensor.

autodidaddict commented:
I'm going to merge this as it's not helping anyone sitting here as a PR. We can always iterate on the interface after this.
